EIP4906: ERC-721 Metadata Update Extension (#4906)
ethereum/EIPs: EIPS/eip-4906.md

The **metadata update extension** is OPTIONAL for ERC-721/ERC-1155 contracts.

```solidity
/// @title ERC-721/ERC-1155 Metadata Update Extension
interface IERC4906  {
    /// @dev This event emits when the metadata of a token is changed.
    /// So that the third-party platforms such as NFT market could
    /// timely update the images and related attributes of the NFT
    event MetadataUpdate(uint256 indexed _tokenId);
}
```

The `MetadataUpdate` event MUST be emitted when the metadata of a token is changed.

## Rationale

Different NFTs have different metadata, and metadata generally has multiple fields. `bytes data` could be used to represents the modified value of metadata.  It is difficult for third-party platforms to identify various types of `bytes data`, so there is only one parameter `uint256 indexed _tokenId` in `MetadataUpdate` event.

After capturing the `MetadataUpdate` event, a third party can update the metadata with information returned from the `tokenURI(uint256 _tokenId)` of ERC721 or `uri(uint256 _tokenId)` of ERC1155.

## Backwards Compatibility

No issues.

> 

Reviewer: There needs to be a Security Considerations section, see [EIP-1](https://eips.ethereum.org/EIPS/eip-1)

---
eip: 4906
title: ERC-721/ERC-1155 Metadata Update Extension
description: Standard interface extension for ERC-721/ERC-115 metadata update 
author: Anders (@0xanders)
discussions-to: https://ethereum-magicians.org/t/eip4906-erc-721-erc-1155-metadata-update-extension/8588
status: Draft
type: Standards Track
category: ERC
created: 2022-03-13

> requires: 721 or 1155

Reviewer 2: ```suggestion
requires: 721, 1155
```

Reviewer 3: Thanks, the document has been modified and committed.

Author: Thanks, the document has been modified and committed.

Reviewer: @MicahZoltu if an EIP uses one of two different EIPS, should they use the notation you suggested? I've been thinking there might be a place for a notation like `721 || 1155, 165` to signify it uses either 721 **or** 1155 (but definitely 165)

Reviewer 2: An implementation of a standard may depend on one of two other standards, but the standard itself still has two dependencies.  This means that in order to understand the full standard, you must understand all of its dependencies (if not, some of the EIP may not make sense).

If you want to keep the dependency chains separate, you can split this EIP into two EIPs, each with a different single dependency.

Author: > An implementation of a standard may depend on one of two other standards, but the standard itself still has two dependencies. This means that in order to understand the full standard, you must understand all of its dependencies (if not, some of the EIP may not make sense).
> 

1. `requires: 721, 1155` has been changed to  `requires: 165, 721`
2. ERC1155 relative description has been removed.

---
eip: 4906
title: ERC-721 Metadata Update Extension
description: Add a `MetadataUpdate` event to ERC-721.
author: Anders (@0xanders), Lance (@LanceSnow)
discussions-to: https://ethereum-magicians.org/t/eip4906-erc-721-erc-1155-metadata-update-extension/8588
status: Draft
type: Standards Track
category: ERC
created: 2022-03-13
requires: 165, 721
---

## Abstract

This standard is an extension of [ERC-721](./eip-721.md). It proposes adding a `MetadataUpdate` event to ERC-721 tokens.

## Motivation

> Many [ERC-721](./eip-721.md) contracts emit a custom event when a token's metadata is changed. It is easy to update metadata of one NFT by specific event, but it is difficult for third-party platforms such as NFT marketplace to update metadata of many NFTs based on custom events.

Reviewer 4: ```suggestion
Many [ERC-721](./eip-721.md) contracts emit an event when one of its token's metadata is changed. While tracking changes based on these different events is possible, it is extra effort for third-party platforms, such as an NFT marketplace, to build bespoke solutions for each NFT collection.
```

> Having a standard `MetadataUpdate` event will make it easy for third-party platforms to timely update metadata of many NFTs.

Reviewer 4: ```suggestion
Having a standard `MetadataUpdate` event will make it easy for third-party platforms to timely update the metadata of many NFTs.
```

## Specification

The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> The **metadata update extension** is OPTIONAL for ERC-721/ERC-1155 contracts.

Reviewer 4: ```suggestion
The **metadata update extension** is OPTIONAL for ERC-721 contracts.
```

```solidity
/// @title ERC-721 Metadata Update Extension
interface IERC4906 is IERC165{
    /// @dev This event emits when the metadata of a token is changed.
    /// So that the third-party platforms such as NFT market could
    /// timely update the images and related attributes of the NFT
    event MetadataUpdate(uint256 _tokenId);
}
```

> The `MetadataUpdate` event MUST be emitted when the metadata of a token is changed.

Reviewer 4: I'd like to see the EIP explicitly state exactly when the event should/shouldn't be emitted. For example, these cases:

 - When a token is minted after the contract is deployed?
 - When a token is minted in the constructor?
 - When a token is burned?
 - When the `tokenURI` changes, but the JSON does not.<- I would assume don't emit?
 - When the `tokenURI` changes, and the JSON also changes. <- I would assume emit
 - When the `tokenURI` does not change, but the JSON does. <- I would also assume emit

Author: > * When a token is minted after the contract is deployed?
> * When a token is minted in the constructor?
> * When a token is burned?
> * When the `tokenURI` changes, but the JSON does not.<- I would assume don't emit?
> * When the `tokenURI` changes, and the JSON also changes. <- I would assume emit
> * When the `tokenURI` does not change, but the JSON does. <- I would also assume emit

The `MetadataUpdate` event MUST be emitted when the JSON metadata of a token is changed.

Not emitting `MetadataUpdate` event is RECOMMENDED when a token is minted.

Not emitting `MetadataUpdate` event is RECOMMENDED  when a token is burned.

Not emitting `MetadataUpdate` event is RECOMMENDED  when the tokenURI changes but the JSON metadata does not.

Reviewer 4: Great, I think that clears up my concern. I think you can put that verbatim into the EIP.

The `supportsInterface` method MUST return `true` when called with `0x49064906`.

## Rationale

Different NFTs have different metadata, and metadata generally has multiple fields. `bytes data` could be used to represents the modified value of metadata.  It is difficult for third-party platforms to identify various types of `bytes data`, so there is only one parameter `uint256 indexed _tokenId` in `MetadataUpdate` event.

> After capturing the `MetadataUpdate` event, a third party can update the metadata with information returned from the `tokenURI(uint256 _tokenId)` of ERC721.

Reviewer 4: ```suggestion
After capturing the `MetadataUpdate` event, a third party can update the metadata with information returned from the `tokenURI(uint256 _tokenId)` of ERC-721.
```

After capturing the `MetadataUpdate` event, a third party can update the metadata with information returned from the `tokenURI(uint256 _tokenId)` of ERC-721.

## Backwards Compatibility

No issues.

## Reference Implementation

```solidity
// SPDX-License-Identifier: CC0
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import "./IERC4906.sol";

contract ERC4906 is ERC721, IERC4906 {

    constructor(string memory name_, string memory symbol_) ERC721(name_, symbol_){
    }

    /// @dev See {IERC165-supportsInterface}.
    function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165, ERC721) returns (bool) {
        return interfaceId == bytes4(0x49064906) || super.supportsInterface(interfaceId);
    }
}
```

## Security Considerations

If there is an off-chain modification of metadata, a method that triggers `MetadataUpdate` can be added, but note the function's permission controls.

## Copyright

> Copyright and related rights waived via [CC0](../LICENSE).

Reviewer 5: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

I had originally missed the .md in my suggestion, sorry!

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 5: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```
